kart/working_copy.py

```diff
@@ -462,3 +463,38 @@ def _check_for_unsupported_structural_changes(
                 "Sorry, this type of filter is not yet supported when there are structural changes."
                 f" Unfilterable structural changes are affecting:\n{unsupported_filters}"
             )
+
+
+def handle_working_copy_tree_mismatch(wc_type_name, actual_tree_id, expected_tree_id):
+    actual_tree_id = f"tree {actual_tree_id}" if actual_tree_id else "the empty tree"
+    expected_tree_id = (
+        f"tree {expected_tree_id}" if expected_tree_id else "the empty tree"
+    )
+
+    message = [
+        f"The {wc_type_name} working copy appears to be out of sync with the repository.",
+        f"The working copy's own records show it is tracking {actual_tree_id};",
+        f"based on the repository it should be tracking {expected_tree_id}.",
+        "The most straight-forward fix is generally to recreate the working copy from the ground up (losing any uncommitted changes in the process.)",
+    ]
+
+    if get_input_mode() != InputMode.INTERACTIVE:
+        message.append("To do so, try the following command:")
+        message.append("kart create-workingcopy --delete-existing --discard-changes")
+        raise BadStateError("\n".join(message), exit_code=BAD_WORKING_COPY_STATE)
+
+    click.echo("\n".join(message))
+    click.echo()
+    if not click.confirm(
+        "Do you want to recreate the working copy from the ground up?"
+    ):
+        raise BadStateError(message[0], exit_code=BAD_WORKING_COPY_STATE)
+
+    from kart.create_workingcopy import create_workingcopy
+
+    click.echo("Recreating working copy ...")
+    ctx = click.get_current_context()
+    subctx = click.Context(ctx.command, parent=ctx)
+    subctx.obj = ctx.obj
+    subctx.invoke(create_workingcopy, delete_existing=True, discard_changes=True)
+    sys.exit(0)
```

Comment:
If you get here by running another command, seems like we should not exit here and instead just run the original command

Reply:
Added a prompt to see if the user still wants to do it. Didn't want to do it by default since it may not still make sense (ie, probably the command depends on the state of the WC in some way or we wouldn't have got here).

Reply:
Can you give an example of what they might be trying to do that would now be a bad thing to do?

status/diff/commit would all be sensible here AFAICT (commit would just noop due to no changes)